Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -140,43 +139,87 @@ namespace ROS2
+    {
+        const bool isIntensityEnabled = m_lidarCore.m_lidarConfiguration.m_lidarSystemFeatures & LidarSystemFeatures::Intensity;
 
         auto* ros2Frame = GetEntity()->FindComponent<ROS2FrameComponent>();
         auto message = sensor_msgs::msg::PointCloud2();
         message.header.frame_id = ros2Frame->GetFrameID().data();
         message.header.stamp = ROS2Interface::Get()->GetROSTimestamp();
         message.height = 1;
-        message.width = lastScanResults.m_points.size();
-        message.point_step = sizeof(AZ::Vector3);
-        message.row_step = message.width * message.point_step;
+        message.width = results.m_points.size();
+
+        sensor_msgs::PointCloud2Modifier modifier(message);
+        if (isIntensityEnabled)
+        {
+            modifier.setPointCloud2Fields(
+                4,
+                "x",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32,
+                "y",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32,
+                "z",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32,
+                "intensity",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32);
+        }
+        else
+        {
+            modifier.setPointCloud2Fields(
+                3,
+                "x",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32,
+                "y",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32,
+                "z",
+                1,
+                sensor_msgs::msg::PointField::FLOAT32);
+        }
+        modifier.resize(results.m_points.size());
+
+        sensor_msgs::PointCloud2Iterator<float> xIt(message, "x");
+        sensor_msgs::PointCloud2Iterator<float> yIt(message, "y");
+        sensor_msgs::PointCloud2Iterator<float> zIt(message, "z");
+
+        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> intensityIt;
+        if (isIntensityEnabled)
+        {
+            intensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+        }
 
-        AZStd::array<const char*, 3> point_field_names = { "x", "y", "z" };
-        for (int i = 0; i < point_field_names.size(); i++)
+        for (size_t i = 0; i < results.m_points.size(); ++i)
         {
-            sensor_msgs::msg::PointField pf;
-            pf.name = point_field_names[i];
-            pf.offset = i * 4;
-            pf.datatype = sensor_msgs::msg::PointField::FLOAT32;
-            pf.count = 1;
-            message.fields.push_back(pf);
+            AZ::Vector3 position = results.m_points[i];
+            *xIt = position.GetX();
+            *yIt = position.GetY();
+            *zIt = position.GetZ();
+
+            if (isIntensityEnabled)
+            {
+                *intensityIt.value() = results.m_intensities[i];
+                ++intensityIt.value();
```

Comment:
This is probably an error - you increment .value() instead of operator itself, like you do with e.g. xIt below.

Reply:
Note that `intenistyIt` is a pointcloud iterator wrapped inside of the optional.